src/util/SignalsHandler.cpp

```diff
@@ -0,0 +1,130 @@
+    ANY  SPECIAL,  DIRECT,  INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include "util/SignalsHandler.hpp"
+
+#include "util/Assert.hpp"
+#include "util/Constants.hpp"
+#include "util/config/Config.hpp"
+#include "util/log/Logger.hpp"
+
+#include <boost/asio/error.hpp>
+#include <boost/asio/executor_work_guard.hpp>
+
+#include <chrono>
+#include <cmath>
+#include <csignal>
+#include <cstddef>
+#include <functional>
+#include <optional>
+#include <string>
+#include <utility>
+
+namespace util {
+
+namespace impl {
+
+class SignalsHandlerStatic {
+    static SignalsHandler* handler_;
+
+public:
+    static void
+    registerHandler(SignalsHandler& handler)
+    {
+        ASSERT(handler_ == nullptr, "There could be only one instance of SignalsHandler");
+        handler_ = &handler;
+    }
+
+    static void
+    resetHandler()
+    {
+        handler_ = nullptr;
+    }
+
+    static void
+    handleSignal(int signal)
+    {
+        ASSERT(handler_ != nullptr, "SignalsHandler is not initialized");
+        handler_->stopHandler_(signal);
+    }
+
+    static void
+    handleSecondSignal(int signal)
+    {
+        ASSERT(handler_ != nullptr, "SignalsHandler is not initialized");
+        handler_->secondSignalHandler_(signal);
+    }
+};
+
+SignalsHandler* SignalsHandlerStatic::handler_ = nullptr;
+
+}  // namespace impl
+
+SignalsHandler::SignalsHandler(Config const& config, std::function<void(std::string)> forceExitHandler)
+    : gracefulPeriod_(0)
+    , context_(1)
+    , stopHandler_([this, forceExitHandler](int) mutable {
+        LOG(LogService::info()) << "Got stop signal. Stopping Clio. Graceful period is "
+                                << std::chrono::duration_cast<std::chrono::milliseconds>(gracefulPeriod_).count()
+                                << " milliseconds.";
+        setHandler(impl::SignalsHandlerStatic::handleSecondSignal);
+        timer_.emplace(context_.scheduleAfter(
+            gracefulPeriod_,
+            [forceExitHandler = std::move(forceExitHandler)](auto&& stopToken) {
+                if (not stopToken.isStopRequested()) {
+                    forceExitHandler("Force exit at the end of graceful period.");
+                }
+            }
+        ));
+        stopSignal_();
+    })
+    , secondSignalHandler_([forceExitHandler = std::move(forceExitHandler)](int) {
+        forceExitHandler("Force exit on second signal.");
+        setHandler();
+    })
+{
+    impl::SignalsHandlerStatic::registerHandler(*this);
+
+    auto const gracefulPeriod =
+        std::round(config.valueOr("graceful_period", 10.f) * static_cast<float>(util::MILLISECONDS_PER_SECOND));
```

Comment:
Do we need to add this to docs and example config?

Reply:
This PR doesn't implement graceful period feature, it is only a part of it. But I added docs and config example about it.